File: projects/cms-flask/App/utils/file_utils/s3_utils.py
from typing import List, Optional
from flask import flash, current_app
import boto3
from io import BytesIO


def ensure_trailing_slash(path: str) -> str:
    return path if path.endswith("/") else path + "/"


def get_s3_client():
    return boto3.client(
        service_name="s3",
        endpoint_url=f"https://{current_app.config['CLOUDFLARE_ID']}.r2.cloudflarestorage.com",
        aws_access_key_id=current_app.config["AWS_ACCESS_KEY_ID"],
        aws_secret_access_key=current_app.config["AWS_SECRET_ACCESS_KEY"],
        region_name="eeur",
    )


def upload_s3_file(file, filepath, app_name):
    s3_client = get_s3_client()
    s3_client.upload_fileobj(file, app_name, filepath)


def get_subdirectories(app_name, prefix=None):
    s3_client = get_s3_client()
    s3_objects = s3_client.list_objects(
        Bucket=app_name,
        Prefix=ensure_trailing_slash(prefix),
        Delimiter="/",
    ).get("CommonPrefixes")
    return [d["Prefix"] for d in s3_objects] if s3_objects else []


def get_subdirectories_and_first_image(app_name, prefix=None):
    s3_client = get_s3_client()
    subdirectories = get_subdirectories(app_name, prefix)
    if not subdirectories:
        return []
    albums = []
    for subdir in subdirectories:
        album_name = subdir.rstrip("/").split("/")[-1]
        images = s3_client.list_objects(Bucket=app_name, Prefix=subdir).get("Contents")
        if not images:
            continue
        first_image_key = images[0]["Key"]
        albums.append(
            {
                "albumName": album_name,
                "firstImageUrl": f"https://bucket.cloud8skate.com/{first_image_key}",
            }
        )
    return albums


def get_filenames(app_name: str, prefix: Optional[str] = None) -> List[str]:
    s3_client = get_s3_client()
    s3_objects = s3_client.list_objects(
        Bucket=app_name,
        Prefix=ensure_trailing_slash(prefix),
        Delimiter="/",
    ).get("Contents")
    return [image["Key"] for image in s3_objects] if s3_objects else []


def get_s3_file(filepath, app_name):
    s3_client = get_s3_client()
    file_obj = BytesIO()
    s3_client.download_fileobj(app_name, filepath, file_obj)
    file_obj.seek(0)
    return file_obj


def delete_directory(prefix, app_name):
    s3_client = get_s3_client()
    response = s3_client.list_objects(
        Bucket=app_name, Prefix=ensure_trailing_slash(prefix)
    )
    if "Contents" not in response:
        flash(
            f"No objects found with prefix {prefix} in bucket {app_name}",
            "warning",
        )
        return
    objects_to_delete = [{"Key": obj["Key"]} for obj in response["Contents"]]
    if objects_to_delete:
        s3_client.delete_objects(
            Bucket=app_name,
            Delete={"Objects": objects_to_delete},
        )
    flash(
        f"All objects with prefix {prefix} have been deleted from bucket {app_name}",
        "success",
    )

File: projects/nx-workspace/apps/ui/small-business-next/src/lib/s3.utils.ts
import {
  S3Client,
  PutObjectCommand,
  ListObjectsV2Command,
  GetObjectCommand,
  DeleteObjectsCommand,
} from '@aws-sdk/client-s3';
import { Readable } from 'stream';
import { getEnvironmentVariable } from '@vigilant-broccoli/common-node';

function ensureTrailingSlash(path: string | undefined): string {
  if (!path) return '';
  return path.endsWith('/') ? path : path + '/';
}

function getS3Client() {
  return new S3Client({
    region: 'auto',
    endpoint: `https://${getEnvironmentVariable('CLOUDFLARE_ACCOUNT_ID')}.r2.cloudflarestorage.com`,
    credentials: {
      accessKeyId: getEnvironmentVariable('AWS_ACCESS_KEY_ID') as string,
      secretAccessKey: getEnvironmentVariable('AWS_SECRET_ACCESS_KEY') as string,
    },
  });
}

export async function uploadS3File(
  file: Buffer | Uint8Array | Blob | string | Readable,
  filepath: string,
  appName: string,
) {
  const s3 = getS3Client();
  await s3.send(
    new PutObjectCommand({
      Bucket: appName,
      Key: filepath,
      Body: file,
    }),
  );
}

export async function getSubdirectories(
  appName: string,
  prefix?: string,
): Promise<string[]> {
  const s3 = getS3Client();
  const res = await s3.send(
    new ListObjectsV2Command({
      Bucket: appName,
      Prefix: ensureTrailingSlash(prefix),
      Delimiter: '/',
    }),
  );
  return res.CommonPrefixes?.map(d => d.Prefix as string) ?? [];
}

export async function getSubdirectoriesAndFirstImage(
  appName: string,
  prefix?: string,
) {
  const s3 = getS3Client();
  const subdirs = await getSubdirectories(appName, prefix);
  const albums: { albumName: string; firstImageUrl: string }[] = [];

  for (const subdir of subdirs) {
    const res = await s3.send(
      new ListObjectsV2Command({
        Bucket: appName,
        Prefix: subdir,
      }),
    );
    const images = res.Contents;
    if (!images || images.length === 0) continue;

    const firstImageKey = images[0].Key as string;
    albums.push({
      albumName: subdir.replace(/\/$/, '').split('/').pop() as string,
      firstImageUrl: `https://bucket.cloud8skate.com/${firstImageKey}`,
    });
  }
  return albums;
}

export async function getFilenames(
  appName: string,
  prefix?: string,
): Promise<string[]> {
  const s3 = getS3Client();
  const res = await s3.send(
    new ListObjectsV2Command({
      Bucket: appName,
      Prefix: ensureTrailingSlash(prefix),
      Delimiter: '/',
    }),
  );
  return res.Contents?.map(img => img.Key as string) ?? [];
}

export async function getS3File(
  filepath: string,
  appName: string,
): Promise<Buffer> {
  const s3 = getS3Client();
  const res = await s3.send(
    new GetObjectCommand({
      Bucket: appName,
      Key: filepath,
    }),
  );
  if (!res.Body) {
    throw new Error('Missing Body');
  }
  return Buffer.from(await res.Body.transformToByteArray());
}

export async function deleteDirectory(prefix: string, appName: string) {
  const s3 = getS3Client();
  const res = await s3.send(
    new ListObjectsV2Command({
      Bucket: appName,
      Prefix: ensureTrailingSlash(prefix),
    }),
  );
  if (!res.Contents || res.Contents.length === 0) return;

  const objectsToDelete = res.Contents.map(obj => ({ Key: obj.Key as string }));
  await s3.send(
    new DeleteObjectsCommand({
      Bucket: appName,
      Delete: { Objects: objectsToDelete },
    }),
  );
}
